h/services/url_migration.py

```diff
@@ -0,0 +1,117 @@
+
+        :param url_mappings: Dict of URL mappings that conform to `URLMigrationSchema`.
+        """
+
+        session = self.request.db
+
+        for url, updates in url_mappings.items():
+            uri_normalized = normalize(url)
+
+            annotations = (
+                session.query(Annotation)
+                .join(DocumentURI, Annotation.document_id == DocumentURI.document_id)
+                .filter(DocumentURI.uri_normalized == uri_normalized)
+            )
+
+            processed_count = 0
+
+            for ann in annotations:
+                ann_update_data = {
+                    "target_uri": updates["url"],
+                }
+
+                if "document" in updates:
+                    ann_update_data["document"] = transform_document(
+                        updates["document"], updates["url"]
+                    )
+
+                # Add selectors to annotation if there is no selector of the same
+                # type.
+                #
+                # This change is specifically to aid in the migration of ebook
+                # annotations from a chapter/page URL to the containing book.
+                # The information about which chapter/page the annotation refers
+                # to is then moved into selectors.
+                #
+                # See https://github.com/hypothesis/h/issues/7709
+                if new_selectors := updates.get("selectors"):
+                    selectors = copy.deepcopy(ann.target_selectors)
+                    for new_sel in new_selectors:
+                        if not any(s for s in selectors if s["type"] == new_sel.type):
+                            selectors.append(new_sel)
+                    ann.target_selectors = selectors
+
+                # Update the annotation's `target_uri` and associated document,
+                # and create `Document*` entities for the new URL if they don't
+                # already exist.
+                h.storage.update_annotation(
+                    self.request,
+                    ann.id,
+                    ann_update_data,
+                    # Don't update "edited" timestamp on annotation cards.
+                    update_timestamp=False,
+                )
+
+                # Create an event that is the same as the one used when the
+                # annotation is updated via the API. We send it to consumers
+                # manually rather than using `request.notify_after_commit`
+                # because that mechanism is not available if we're being called
+                # from a Celery task. Also event handlers in `h.subscribers`
+                # try to start a new transaction, which we don't want.
+                event = AnnotationEvent(self.request, ann.id, "update")
+
+                # Re-index annotation in Elasticsearch.
+                search_index = self.request.find_service(name="search_index")
+                search_index.handle_annotation_event(event)
```

Comment:
I can see why the convoluted-ness of the code around `SearchIndexService`, `storage.py`, etc is confusing but I don't think you want to call `search_index.handle_annotation_event()` here.

The call to `storage.update_annotation()` above will add the annotations to the job queue (a.k.a "asynchronous reindexing") which means they will get re-indexed in Elasticsearch as soon as the periodic reindexing task can process those jobs. The job queue/asynchronous reindexing is the appropriate thing to use here: it's designed to be able to [handle large batches of annotations at once from admin pages](https://docs.google.com/presentation/d/1msC6H5wxGQUo80zUWIe1rmfAAoQtUqIlrLFdBCMD-O4/edit#slide=id.ga6bcfe4558_1_34).

`search_index.handle_annotation_event()` is part of h's [synchronous annotation indexing](https://docs.google.com/presentation/d/1msC6H5wxGQUo80zUWIe1rmfAAoQtUqIlrLFdBCMD-O4/edit#slide=id.gc6fa3c898_0_28) which is meant to be used by the annotation create/update APIs. This is **not** designed to be able to handle large batches being added at once. I don't think you want to call this here.

See also: notes on [how and why h combines synchronous and asynchronous reindexing](https://docs.google.com/presentation/d/1msC6H5wxGQUo80zUWIe1rmfAAoQtUqIlrLFdBCMD-O4/edit#slide=id.g9f11104581_0_114)